### PR TITLE
ADR for Kryo class registration policy

### PR DESCRIPTION
**Pull request by tbekas**

### CODE REVIEW

**Review by marcinwadon — DISMISSED** [review]
I like that idea :+1:

File: doc/adr/0009-kryo-class-registration-policy.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+## Context
+
+Having reliable serialization across different modules require using consistent and 
+non-conflicting registration ids.
+
+## Decision
+
+* Each module containing classes that are a subject of a kryo serialization declares 
+its kryo registrar - a public constant of type `Map[Class[_], Int]` representing kryo
```
Comment (kpudlik) [inline doc/adr/0009-kryo-class-registration-policy.md:17]: As a bonus feature, we can check if `Class[_]` have duplicates and if `Int` have duplicates and throw if it happens as we talked.

File: doc/adr/0009-kryo-class-registration-policy.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+| module     | min id | max id |
+| ---------- | ------ | ------ |
+| shared     | 100    | 299    |
+| kernel     | 300    | 399    |
+| sdk        | 400    | 499    |
+| dag-shared | 500    | 599    |
+| core       | 600    | 699    |
+| dag-l1     | 700    | 799    |
+
+* State-channels can use registration ids starting from 1000.
```
Comment (kpudlik) [inline doc/adr/0009-kryo-class-registration-policy.md:35]: That one as well can be secured by refined type

File: doc/adr/0009-kryo-class-registration-policy.md (hunk 1 above)
Comment (tbekas) [reply doc/adr/0009-kryo-class-registration-policy.md:17]: Fully agree, that's an implementation detail however I think so I would like to not include it in ADR.

File: doc/adr/0009-kryo-class-registration-policy.md (hunk 2 above)
Comment (tbekas) [reply doc/adr/0009-kryo-class-registration-policy.md:35]: It actually already is. There's a type alias `KryoRegistrationId = Int Refined Greater[1000]` in `org.tesselation.kernel`.

File: doc/adr/0009-kryo-class-registration-policy.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+| module     | min id | max id |
+| ---------- | ------ | ------ |
+| shared     | 100    | 299    |
+| kernel     | 300    | 399    |
+| sdk        | 400    | 499    |
+| dag-shared | 500    | 599    |
+| core       | 600    | 699    |
+| dag-l1     | 700    | 799    |
+
```
Comment (TheMMaciek) [inline doc/adr/0009-kryo-class-registration-policy.md:34]: Is the range [800, 999] intended as a backup?

Comment (tbekas) [reply doc/adr/0009-kryo-class-registration-policy.md:34]: Yes, it's for future use.

**Review by TheMMaciek — APPROVED** [review]
Sounds good!

**Review by kpudlik — APPROVED** [review]

**Pull request merged**